## Statistics 207  Exam 2
---

## Name: Michael Cao
---

## Net ID: minhcao2
---

#### Release Date: Monday, April 13, 2020

#### Due Date: Committed and pushed before 11:59pm  Tuesday, April 14, 2020

Scoring: Each Problem section is worth 5 points, for a total of 100 points on the exam.

Do all your work in this notebook. Show your work unless directed only to provide the answer. This is an open notes exam. You must work independently  with no assistance from anyone else.

---


### Grader Scoring

Problem | Number of parts | Points possible | Your Score
--- | --- | --- | ---
1 | 4 | 20 | $\qquad$
2 | 4 | 20 | $\qquad$
3 | 4 | 20 | $\qquad$
4 | 4 | 20 | $\qquad$
5 | 4 | 20 | $\qquad$
Total | 20 | 100 | $\qquad\qquad$

---
**Problem 1.** 

A polling organization conducted a statewide survey concerning the race for governor. In a random phone number survey, 400 registered voters responded. It was found that 230 would vote for the Democratic candidate. Let p represent the proportion of registered voters in the state that would vote for the Democratic candidate. Assume the sample of respondents is equivalent to a random sample from the population of registered voters in the state.

**a)** Calculate an estimate for $p$ based on the poll results.

In [3]:
p = 230/400
p

0.575

**b)** Calculate the standard error for the estimate of $p$.

In [5]:
SE = np.sqrt((p * (1 - p)) / 400)
SE

0.024717149916606486

In [9]:
import numpy as np
from scipy.stats import norm

**c)** We wish to test the null hypothesis $H_0: p=0.5$ against the alternative $H_A: p\ne 0.5$. Based on the poll results, calculate the p-value of a large sample z-test of $H_0$.

In [20]:
z = (0.575 - 0.5)/ SE
z
p_value = 2*(1 - norm.cdf(z))
z, p_value

(3.0343304245450398, 0.0024107023750756795)

**d)** The odds of an event is the ratio of the probability that the event occurs to the probability of that the event does not occur. Using your results above, compute an estimate of the odds that a randomly chosen registered voter would vote for the Democrative candidate.

In [22]:
odd = p / (1 - p)
odd

1.352941176470588

---
**Problem 2.** 

Data were collected from 39 students in a university class. The variables were:

+ happy: Happiness on a 10 point scale, where 10 is most happy

+ money: family income in thousands of dollars

+ sex: 1 = satisfactory sexual activity, 0 = not

+ love: 1 = lonely, 2 = secure relationships, 3 = deep feeling of belonging and caring

+ work: 5 point scale where 1 = no job, 3 = OK job, 5 = great job

The data are in the included file 'happy.csv'.

**a)** We read the data into a data frame and display the correlation matrix of all the variables.

In [24]:
# keep
import pandas as pd
df2 = pd.read_csv('happy.csv')
df2.corr()

happy     money       sex      love      work
happy  1.000000  0.270942 -0.032590  0.784219  0.539170
money  0.270942  1.000000  0.306506  0.125815  0.067748
sex   -0.032590  0.306506  1.000000  0.047160 -0.316343
love   0.784219  0.125815  0.047160  1.000000  0.385696
work   0.539170  0.067748 -0.316343  0.385696  1.000000

We are interested in which of the variables are correlated with happiness, either positively or negatively. Which of the other variables have a sample correlation with 'happy' that is larger than 0.50 in absolute value? List all that do.

love, work

**b)** Below we fit a linear regression model with 'happy' as the response variable and 'money', 'sex', 'love and 'work' as explanatory variables. Display the model summary including at least the R-square value, the F-statistic for the model, the model coefficient estimates, and the coefficient standard errors.

In [25]:
# Keep
import statsmodels.api as sm
import statsmodels.formula.api as smf
model1 = smf.ols('happy ~ money + sex + love + work', data=df2).fit()

In [26]:
model1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  happy   R-squared:                       0.710
Model:                            OLS   Adj. R-squared:                  0.676
Method:                 Least Squares   F-statistic:                     20.83
Date:                Mon, 13 Apr 2020   Prob (F-statistic):           9.36e-09
Time:                        11:06:52   Log-Likelihood:                -54.877
No. Observations:                  39   AIC:                             119.8
Df Residuals:                      34   BIC:                             128.1
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.0721      0.853     -0.085      0.933      -1.805       1.660
money          0.0096      0.005      1.837      0.075      -0.001       0.020
sex           -0.1490      0.419     -0.356      0.724      -1.000       0.702
love           1.9193      0.295      6.496      0.000       1.319       2.520
work           0.4761      0.199      2.388      0.023       0.071       0.881
==============================================================================
Omnibus:                        0.561   Durbin-Watson:                   1.541
Prob(Omnibus):                  0.755   Jarque-Bera (JB):                0.170
Skew:                          -0.155   Prob(JB):                        0.918
Kurtosis:                       3.093   Cond. No.                         373.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**c)** List all of the explanatory variables whose coefficients estimates are significantly different than zero at an alpha significance level of 0.05.

Love and Work

**d)** The regression model has the form:

$$ 
E(\text{happy}) = b_0 + b_1 * \text{money} + b_2 * \text{sex} + b_3 * \text{love} + b_4 * \text{work}.
$$

We perform the following F test:

In [27]:
# keep
model1 = smf.ols('happy ~ money + sex + love + work', data=df2).fit()
model0 = smf.ols('happy ~ love + work', data=df2).fit()
f, p, df = model1.compare_f_test(model0)
pd.DataFrame({'f': [f], 'pvalue': [p], 'df_diff': [df]})

f    pvalue  df_diff
0  1.725173  0.193371      2.0

State the null hypothesis for this F test in terms of the population model coefficients.

$H_0$: $B_1$ = $B_2$ = 0

---
**Problem 3.** 

The drug AZT was developed to try to delay the onset of AIDS in HIV positive patients. In a randomized double-blind study 950 HIV positive volunteers without AIDS were assigned to receive either AZT or placebo. At the end of the study the results were as follows.

| $\quad$ | No Aids | Aids |
|---|---|---|
| AZT | 413 | 12 |
| Placebo | 392 | 33 |



**a)** For patients on AZT, calculate the estimated odds of developing AIDS over a time period of the same length as in the study. Do the same for patients on Placebo. Display and label the two odds estimates.

In [28]:
p_of_aids_with_AZT = 12 / (12 + 413)
p_of_aids_without_AZT = 33 / (33 + 392)
p_of_aids_with_AZT, p_of_aids_without_AZT

(0.02823529411764706, 0.07764705882352942)

In [29]:
odd_of_aids_with_AZT = p_of_aids_with_AZT / (1 - p_of_aids_with_AZT)
odd_of_aids_without_AZT = p_of_aids_without_AZT / (1 - p_of_aids_without_AZT)
odd_of_aids_with_AZT, odd_of_aids_without_AZT

(0.029055690072639227, 0.08418367346938777)

**b)** Calculate an odds ratio estimate to compare the effectiveness of AZT to Placebo in delaying the onset of AIDS. Based on your odds ratio, which treatment appears more effective and why?

In [31]:
odd_of_aids_without_AZT / odd_of_aids_with_AZT

2.897321428571429

In [35]:
odd_ratio = (413/392) / (12/33)
odd_ratio

2.8973214285714284

Treatment with AZT is more effective because the odd ratio is > 1

**c)** Calculate the natural logarithm of the odds ratio you computed in b), and compute the standard error of the log-odds-ratio.

In [36]:
log_odd_ratio = np.log(odd_ratio)
log_odd_ratio

1.0637866648490502

In [37]:
SE_log_odd_ratio = np.sqrt(1/413 + 1/392 + 1/12 + 1/33)
SE_log_odd_ratio

0.3443961259227231

**d)** Under the null hypothesis that there is no difference in effectiveness between AZT and Placebo, what is the hypothesized value of the population log-odds-ratio? 

In [42]:
z = odd_ratio/SE_log_odd_ratio
z


8.412758479233126

We can conclude that the AZT drug is effective because log-odds-ratio is not 0, also, z value for odd ratio is very high : 8.412, hence, we can reject null hypothesis

---
**Problem 4.**

Data were collected from 39 students in a university class. Some of the variables were:

+ happy: Happiness on a 10 point scale, where 10 is most happy

+ money: family income in thousands of dollars

+ sex: 1 = satisfactory sexual activity, 0 = not

+ love: 1 = lonely, 2 = secure relationships, 3 = deep feeling of belonging and caring

+ work: 5 point scale where 1 = no job, 3 = OK job, 5 = great job

The data are in the included file 'happy.csv'.

Consider the following Python code and results:

In [43]:
# keep
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf

In [47]:
# keep
df4 = pd.read_csv("happy.csv")
logit1 = smf.logit('sex ~ love + work', data=df4).fit()
print(logit1.summary())

Optimization terminated successfully.
         Current function value: 0.540229
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:                    sex   No. Observations:                   39
Model:                          Logit   Df Residuals:                       36
Method:                           MLE   Df Model:                            2
Date:                Mon, 13 Apr 2020   Pseudo R-squ.:                  0.1248
Time:                        11:42:10   Log-Likelihood:                -21.069
converged:                       True   LL-Null:                       -24.072
Covariance Type:            nonrobust   LLR p-value:                   0.04961
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.5938      1.899      1.366      0.172      -1.128       6.315
love           0.8519      0.

**a)** For the individual with the record shown below, find the predicted probability for 'sex'=1 based on the fitted model.

        happy     6
        money    45
        sex       0
        love      2
        work      3

In [51]:
y_hat = 2.5938 + 0.8519*2 - 1.0997*3
y_hat

0.9985000000000004

**b)** Which, if any, of the explanatory variables has a coefficient that is significantly different from zero at an alpha level of 0.05?

Work

**c)** Below is the confusion matrix if the model 'logit1' is used as a predictive classifier: we predict 'sex=1' for fitted probabilities $\ge 0.50$, and we predict 'sex=0' for fitted probabilities $\lt 0.50$.

In [53]:
# keep
from sklearn.metrics import confusion_matrix, roc_curve, roc_auc_score
yhat = 1*(logit1.fittedvalues >= 0.50)
tn, fp, fn, tp = confusion_matrix(y_true=df4['sex'], y_pred=yhat).ravel()
pd.DataFrame({'Actual 0': [tn, fp], 
              'Actual 1': [fn, tp] }, 
             index=['Model Classified as 0', 'Model Classified as 1'])

Actual 0  Actual 1
Model Classified as 0         6         4
Model Classified as 1         6        23

If the sensitivity of the classifier is the probability of classifying as 1 given that the true value is 1, what is the estimated sensitivity implied by the confusion matrix shown here?

In [56]:
23/(23+4)

0.8518518518518519

**d)** If the specificity of a classifier is the probability of classifying as 0 given that the true value is zero, find the estimated specificity of the classifer based on the model 'logit1'.

In [83]:
logit1.params

Intercept    2.593792
love         0.851946
work        -1.099733
dtype: float64

In [88]:
def senspec(y, score, thresh):
    yhat = 1*(score >= thresh)
    tn, fp, fn, tp = confusion_matrix(y_true=y, y_pred=yhat).ravel()
    sens = tp / (fn + tp)
    spec = tn / (fp + tn)
    return pd.DataFrame({'tn':[tn], 
                         'fp':[fp], 
                         'fn':[fn], 
                         'tp':[tp], 
                         'sens':[sens], 
                         'spec':[spec]})

In [90]:
phat = logit1.predict(exog=df4[['work', 'love']])
phat.head(10)

0    0.679330
1    0.982874
2    0.413616
3    0.864177
4    0.776654
5    0.679330
6    0.679330
7    0.730760
8    0.890730
9    0.278243
dtype: float64

In [102]:
senspec(df4['sex'], phat, 0.5)

tn  fp  fn  tp      sens  spec
0   6   6   4  23  0.851852   0.5

In [103]:
speci = 6 / 12 
speci

0.5

---
**Problem 5.**

Data were collected on a sample of iris flowers of three different species. The data are in the file 'iris.csv'. The data are read in below and the variable names are displayed.

In [57]:
# keep
df5 = pd.read_csv('iris.csv')
df5.columns

Index(['SepalLength', 'SepalWidth', 'PetalLength', 'PetalWidth', 'Species'], dtype='object')

In [58]:
df5

SepalLength  SepalWidth  PetalLength  PetalWidth    Species
0            5.1         3.5          1.4         0.2     setosa
1            4.9         3.0          1.4         0.2     setosa
2            4.7         3.2          1.3         0.2     setosa
3            4.6         3.1          1.5         0.2     setosa
4            5.0         3.6          1.4         0.2     setosa
..           ...         ...          ...         ...        ...
145          6.7         3.0          5.2         2.3  virginica
146          6.3         2.5          5.0         1.9  virginica
147          6.5         3.0          5.2         2.0  virginica
148          6.2         3.4          5.4         2.3  virginica
149          5.9         3.0          5.1         1.8  virginica

[150 rows x 5 columns]

**a)** Using the pandas groupby function, or otherwise, compute the sample means of 'PetalLength' for each of the three species.

In [80]:

dfnew = df5.groupby("Species").mean()
dfnew

SepalLength  SepalWidth  PetalLength  PetalWidth
Species                                                     
setosa            5.006       3.428        1.462       0.246
versicolor        5.936       2.770        4.260       1.326
virginica         6.588       2.974        5.552       2.026

**b)** Fit a oneway anova regression model with 'PetalLength' as the response and 'Species' as the categorical explanatory variable. Display the model summary.

In [69]:
mod = smf.ols("SepalLength ~ Species", data = df5).fit()
mod.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:            SepalLength   R-squared:                       0.619
Model:                            OLS   Adj. R-squared:                  0.614
Method:                 Least Squares   F-statistic:                     119.3
Date:                Mon, 13 Apr 2020   Prob (F-statistic):           1.67e-31
Time:                        11:59:53   Log-Likelihood:                -111.73
No. Observations:                 150   AIC:                             229.5
Df Residuals:                     147   BIC:                             238.5
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=========================================================================================
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
Intercept                 5.0060      0.073     68.762      0.000       4.862       5.150
Species[T.versicolor]     0.9300      0.103      9.033      0.000       0.727       1.133
Species[T.virginica]      1.5820      0.103     15.366      0.000       1.379       1.785
==============================================================================
Omnibus:                        1.188   Durbin-Watson:                   2.043
Prob(Omnibus):                  0.552   Jarque-Bera (JB):                0.785
Skew:                           0.119   Prob(JB):                        0.675
Kurtosis:                       3.263   Cond. No.                         3.73
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**c)** We would like to test the following null hypothesis at level $\alpha=0.01$:

$$
H_0: \mu_{\text{setosa}} = \mu_{\text{versicolor}} = \mu_{\text{virginica}},
$$

against the alternative that at least one pair of means is different. Here each $\mu_{\text{species}}$ represents the population mean petal length for that species.

Explain what results in the summary table provide a test of this hypothesis, and state what your conclusion is. Does the test reject $H_0$?

The F test is highly significant, F = 119.3, p value < 0.01 => reject null hypothesis

**d)** Determine the estimated mean difference in petal length between virginica and setosa, and give a 95% confidence interval for the difference.

In [74]:
lower_bound, upper_bound = 1.379, 1.785
lower_bound, upper_bound

(1.379, 1.785)

In [75]:
mean_difference = 6.588-5.006
mean_difference

1.5819999999999999